provider/datagen/src/transform/collator/transform.rs

```diff
@@ -0,0 +1,339 @@
+            //     let data = $provider::load_data(root_dir).map_err(|e| {
+            //         DataError::custom("Could not create provider").with_display_context(&e)
+            //     })?;
+            //     Ok(Self { data })
+            // }
+
+            fn load_data(root_dir: &Path) -> Result<HashMap<String, $serde_struct>, DatagenError> {
+                let mut result = HashMap::new();
+                for path in get_dir_contents(&root_dir)? {
+                    let stem_bytes = if let Some(stem_bytes) = path
+                        .file_stem()
+                        .and_then(|p| p.to_str())
+                        .ok_or_else(|| DatagenError::Custom("Invalid file name".to_string(), None))?
+                        .as_bytes()
+                        .strip_suffix($suffix)
+                    {
+                        stem_bytes
+                    } else {
+                        continue;
+                    };
+                    let mut key = String::from_utf8(stem_bytes.to_vec()).map_err(|_| {
+                        DatagenError::Custom("Non-UTF-8 file name".to_string(), None)
+                    })?;
+                    let toml_str = read_path_to_string(&path)?;
+                    let toml_obj: $serde_struct = toml::from_str(&toml_str)
+                        .map_err(|e| DatagenError::Custom(e.to_string(), None))?;
+                    key.make_ascii_lowercase();
+                    result.insert(key, toml_obj);
+                }
+                Ok(result)
+            }
+        }
+
+        impl ResourceProvider<$marker> for $provider {
+            fn load_resource(&self, req: &DataRequest) -> Result<DataResponse<$marker>, DataError> {
+                if self.data.read().unwrap().is_none() {
+                    let root_dir = self.source.get_coll_root()?;
+                    let data = $provider::load_data(root_dir).map_err(|e| {
+                        DataError::custom("Could not create provider").with_display_context(&e)
+                    })?;
+                    *self.data.write().unwrap() = Some(data);
+                }
+
+                let langid = req.options.get_langid();
+                let mut s = String::new();
+                langid.write_to(&mut s).map_err(|_| {
+                    DataErrorKind::MissingResourceOptions.with_req($marker::KEY, req)
+                })?;
+                if &s == "und" {
+                    s = String::from("root");
+                } else {
+                    // No safe method for in-place replacement.
+                    s = s.replace('-', "_");
+                    s.make_ascii_lowercase();
+                }
+                if let Some(extension) = &req.options.get_unicode_ext(&unicode_ext_key!("co")) {
+                    let extension_string = extension.to_string();
+                    let extension_str = &extension_string[..];
+                    s.push('_');
+                    match extension_str {
```

Comment:
```suggestion
                    s.push('_');
                    match &extension.to_string() {
```

Reply:
What would you put in the `_` case in order not to stringify `extension` twice?

Reply:
ah, you can just match it with `s`

Reply:
I guess not `s` because you already have an `s`, but any variable name

Reply:
I omitted the named slice.